src/dune_engine/build_config.ml

```diff
@@ -17,12 +17,41 @@ module Gen_rules = struct
     let union a b = Path.Build.Map.union a b ~f:(fun _ a b -> Some (Subdir_set.union a b))
   end
 
+  module Directory_targets = struct
+    include Build_config_intf.Directory_targets
+
+    let empty =
+      { all = Memo.return Path.Build.Map.empty; mem = (fun _ -> Memo.return false) }
+    ;;
+
+    let of_map all =
+      { all = Memo.return all; mem = (fun p -> Memo.return (Path.Build.Map.mem all p)) }
+    ;;
+
+    let union_exn t { all; mem } =
+      let open Memo.O in
+      let all =
+        Memo.Lazy.create (fun () ->
+          let+ left = t.all
+          and+ right = all in
+          Path.Build.Map.union_exn left right)
+      in
+      let mem p =
+        t.mem p
+        >>= function
+        | true -> Memo.return true
+        | false -> mem p
+      in
+      { all = Memo.Lazy.force all; mem }
+    ;;
+  end
```

Comment:
Just so I understand correctly: we only ever construct this through `of_map` and `union_exn`, right? So what this basically does it provide O(1) union?

If so, then it might be clearer to just have a generic data structure which does just that. One thing I'm worried about is that if we union multiple `Directory_targets` and then evaluate `all`, we will hang on to both the materialised union (in `all`) and the individual bits (in `mem`), effectively doubling the memory requirements. Here is a sketch (not type checked) of a data structure which provides O(1) union but doesn't have that problem:

```
type ('a,'b) t = { all : ('a,'b) Map.t Lazy.t; mem : ('a -> bool) ref }

let union_exn t u =
  let mem = ref (fun x -> !t.mem x || !u.mem x) in
  let all = lazy (
    let all = Map.union_exn (force t.all) (force u.all) in
    mem := (fun x -> Map.mem all x);
    all)
  in
  { all; mem }
```

Another potential issue is that `mem` now does a linear search over the unioned components where before, it was O(log N). Do you expect that to matter in practice?